src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -780,6 +789,27 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
         continue;
       }
 
+      // At this point we know we've hit a match. Error out if
+      //   1) The target assembly is Datadog.Trace.ClrProfiler.Managed
+      //   2) The calling assembly is domain-neutral
+      //   3) Datadog.Trace.ClrProfiler.Managed was never loaded domain-neutral
+      //        or it more recently has been loaded in a named AppDomain
+      // is domain-neutral but Datadog.Trace.ClrProfiler.Managed is
+      if (method_replacement.wrapper_method.assembly.name == "Datadog.Trace.ClrProfiler.Managed"_W &&
+          runtime_information_.is_desktop() && corlib_module_loaded &&
+          module_metadata->app_domain_id == corlib_app_domain_id &&
+          (!managed_profiler_loaded_domain_neutral || managed_profiler_unsafe_to_instrument_domain_neutral)) {
```

Comment:
Interesting that you added this: `method_replacement.wrapper_method.assembly.name == "Datadog.Trace.ClrProfiler.Managed"_W`
Is that necessary because of the `InterceptMethod` reference in the Loader assembly?

Might be a micro-optimization, but I think that's important at this level: Would you be amenable to re-ordering these conditions so that the string comparison is last?
Do you think that would make much of a difference here?

Reply:
Good idea. Yes, I will re-order the conditions so the string comparison is done last.